### ISSUE TITLE
Async proofs: A way to eliminate the need for "Proof using ..." annotations?

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#16266]
`Proof using ...` annotations are useful but awkward for users.  While it's not too hard to create tooling that will [add the necessary annotations automatically](https://github.com/coq/coq/pull/16177), removing incorrect annotations automatically is harder/messier, and you probably have to do an extra full compile to update the annotations.

The manual describes sections primarily as naming scopes, provided for user convenience, so I'd expect that every script containing sections could be rewritten without sections.  Is that true?

If there are no sections in a script, then `Proof using` isn't needed.

Is it feasible for Coq to internally assign fresh names of global scope to theorems and variables defined in sections (in effect making `Proof using` unnecessary) and then later substitute the user-visible short names as part of `Qed` checking (to get the current user-visible behavior for sections).

Does this make any sense?

**Comment by Blaisorblade** [comment]
1. Current `Proof using` annotations are necessary when `Qed` checking does _not_ happen (as in `vos` proofs), so any proposal to infer `Proof using` must address that. Deducing the optimal `Proof using` is easy when `Qed` checking is done, and impossible otherwise.

> The manual describes sections primarily as naming scopes, provided for user convenience, so I'd expect that every script containing sections could be rewritten without sections. Is that true?

2. Picking names is not enough to "rewrite sections away". Take `Lemma foo : ... Proof using A B.`. `Proof using` annotations must be completed (as explained for `Proof using Type`), so say this completes to `Proof using A B C.`.
_After the section ends_, theorems take _some_ variables as section arguments. Without `Proof using`, they only take the variables they use. With this `Proof using` annotation, the theorem takes as extra args exactly section variables `A`, `B` and `C` , and nothing else.

**Comment by jfehrle** [comment]
OK, I'll consider that a dead end.  Maybe the best we can do is to add tooling that automatically removes incorrect annotations _en masse_ rather than quitting at the first error (since my simple tooling runs in coqc).

That still requires a full compile to update the annotations, but at least it would be automated.  Perhaps later on Coq could process a proof, discover that it's not annotated optimally, change the annotations and re-process the proof.

**Comment by mrhaandi** [comment]
I was trying to get rid of Sections in a larger project by using Functors. Sections seem to be the most convenient way to create functions `A1 -> .. -> An -> B` with the necessary assumptions, and also work in `vos` mode with `Proof using` annotations.
On the other hand, Functors just create `B` with Axioms `A1 .. An` and I could not extract any function.
For my use case, it would have helped a lot to a-posteriori turn a definition of type `B` which uses an Axiom `A` into a function `A -> B`. In fact, if this were possible, then Functors would be more convenient than Sections because they can be combined.

**Comment by ejgallego** [comment]
I have been defending this for years, IMO the information Proof Using is capturing should be stored in the cache of the document manager, we recently did some more progress on a new document manager and indeed it'd be easy to do so I think.

**Comment by jfehrle** [comment]
The important thing for annotations is that they persist across Coq processes, either in the .v files (currently) or in some other (new?) file type.  Otherwise you only get async speedup in the GUI the second time a proof is processed and never in coqc.  Is the document manager cache persistent?  If so, would you check its contents into git?  (Also, I don't know what functionality the document manager is supposed to provide.)

For now, I can add an option to my prototype that automatically removes all `using` annotations.  Users can run that, then do a second run to automatically add annotations.  I expect the removals can be done with `-vos`, so it would much faster than a regular compile.  Not would it generate failures upon `Qed` typechecking since it's not done in that mode.  While not ideal, I think that's still a useful improvement--much easier than updating the annotations by hand.

**Comment by Blaisorblade** [comment]
IIUC today they're saved in `.aux` files, but I don't understand well how those are meant to be updated soundly.

> I expect the removals can be done with -vos, so it would much faster than a regular compile

But with `-vos` you don't have a proof body so you don't know which section variables are used in the body?

**Comment by ejgallego** [comment]
Yes, the cache of the document manager is meant to be persistent, but it is a binary cache essentially so not sure git is the best way to share it.

**Comment by jfehrle** [comment]
> IIUC today they're saved in `.aux` files, but I don't understand well how those are meant to be updated soundly.

Based on a very quick look, I think that's not the case.  It looks like timing information only.  See `aux_file.ml`.  Below is a truncated sample (filename is `.auxiliary.aux`, note the dot at the beginning).  Moving the info from the .v files raises more implementation issues, e.g. migration/backward compatibility, more than I'd want to take on for a quick but imperfect improvement.  I think the long-term best solution is probably allowing the compiler to reprocess proofs when necessary.

```
COQAUX1 1eb1dbab9b3558e9c3c1d0442d697e90 /home/proj/coq2/_build/default/theories/ZArith/auxiliary.v
0 0 VernacProof "tac:no using:no"
1184 1188 proof_build_time "0.012"
0 0 Zne_left "0.012"
1153 1183 context_used ""
1184 1188 proof_check_time "0.001"
0 0 VernacProof "tac:no using:no"
1267 1271 proof_build_time "0.001"
0 0 Zegal_left "0.001"
```

> But with `-vos` you don't have a proof body so you don't know which section variables are used in the body?

Correct.  That's why you need two runs to update everything.  You don't need to know which section variables are used to remove all the `using` annotations.

**Comment by Blaisorblade** [comment]
Ah, you're removing _all_ `using` annotations.

> It looks like timing information only.

I recall @gares talking of this feature, and `context_used` and code around it (in https://github.com/coq/coq/blob/d1245c6f48b0349045f8758948a09ae112c0b867/vernac/declare.ml#L337) seems relevant...

I can also see it being used e.g. for https://gitlab.mpi-sws.org/iris/stdpp/-/blob/45fd520d555e9d9bfa3d3894ae66079fbbce17f0/theories/fin_maps.v. An extract I had around:
```
COQAUX1 07dc136845dd9be4ab2a9085c4cd06c9 /Users/pgiarrusso1/git-bedrock/iris/_build/default/stdpp/theories/fin_maps.v
0 0 VernacProof "tac:no using:no"
8856 8860 proof_build_time "0.001"
0 0 map_eq_iff "0.001"
8842 8855 context_used "H6 EqDecision0 H5 H4 H3 H2 H1 H"
8856 8860 proof_check_time "0.000"
0 0 VernacProof "tac:no using:no"
9111 9115 proof_build_time "0.017"
0 0 map_subseteq_spec "0.017"
9023 9110 context_used ""
```

**Comment by gares** [comment]
yup. that info is only used in interactive mode or quick mode iif the hash of the file, first line, did not change.
